### DISCUSSION TITLE
Signac scATAC-seq data integration Error

### DISCUSSION THREAD

**Discussion by biobarani** [stuart-lab/signac#1009] [General]
Hello Authors,
thanks for this wonderful program.

While running scATAC-seq data integration vignette, I'm getting error for the following step
# quantify multiome peaks in the scATAC-seq dataset
counts <- FeatureMatrix(
  fragments = atac.frags,
  features = granges(Aggr_Cd8a_pos),
  cells = atac.cells
)

Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘granges’ for signature ‘"Assay"’

I'm not sure how to get rid of this error, please help in this. Thanks in advance.

**Comment by timoast** [comment]
You might need to change the default assay for `Aggr_Cd8a_pos` if it does not have a chromatin assay as the active assay. You can use the `DefaultAssay()` function to do this. Alternatively, you can extract the granges for a specific assay using the ChromatinAssay method, eg: `granges(Aggr_Cd8a_pos[['assay']])` (replacing 'assay' with the name of your chromatin assay in the object.

**Reply by biobarani** [reply]
Hi Timoast,
Thank you very much for your very quick reply.
Sorry, I'm still not able to rectify this error, I have changed those. For your better clarity, I have attached both RNA preprocessed script and scATACseq integration script here, Kindly help me to sort out this error. Thanks.


library(Signac)
library(Seurat)
library(ggplot2)


Aggr_Cd8a_pos <- Read10X("~/Desktop/outs/filtered_feature_bc_matrix/")

######## I initially run the RNAseq part using standard Seurat pipeline to get preprocessed rds file.

### analysis
Aggr_Cd8a_pos <- NormalizeData(Aggr_Cd8a_pos)
Aggr_Cd8a_pos <- ScaleData(Aggr_Cd8a_pos)
Aggr_Cd8a_pos <- FindVariableFeatures(Aggr_Cd8a_pos)
Aggr_Cd8a_pos <- RunPCA(Aggr_Cd8a_pos, npcs = 30, verbose = F)
plot(Aggr_Cd8a_pos@reductions$pca@stdev)

Aggr_Cd8a_pos <- RunTSNE(Aggr_Cd8a_pos, dims = 1:20, perplexity = 100)
Aggr_Cd8a_pos <- RunUMAP(Aggr_Cd8a_pos, dims = 1:20, perplexity = 100)

##### Check gene expression
DefaultAssay(Aggr_Cd8a_pos) <- "RNA"

##### Clustering
Aggr_Cd8a_pos <- FindNeighbors(Aggr_Cd8a_pos, dims = 1:30)
Aggr_Cd8a_pos <- FindClusters(Aggr_Cd8a_pos, resolution = 0.3)
Aggr_Cd8a_pos$cluster <- Aggr_Cd8a_pos$RNA_snn_res.0.3
DimPlot(Aggr_Cd8a_pos, reduction = "umap",pt.size = 1,split.by = "orig.ident")+ggtitle("4samples_multiome_aggr Cd8a+ cells")

saveRDS(Aggr_Cd8a_pos,"~/Desktop/output/Aggr_Cd8a_pos.rds") 

library(Signac)
library(Seurat)
library(ggplot2)

### load the pre-processed multiome data
Aggr_Cd8a_pos <- readRDS("~/Desktop/output/Aggr_Cd8a_pos.rds")

### process the scATAC data ( to avoid confusion I used the same variable names which are used in the vignette)
pbmc.multi  <- Aggr_Cd8a_pos


### first count fragments per cell
fragpath <- "~/Desktop/output/atac_fragments.tsv.gz"
fragcounts <- CountFragments(fragments = fragpath)
atac.cells <- fragcounts[fragcounts$frequency_count > 2000, "CB"]

###create the fragment object
atac.frags <- CreateFragmentObject(path = fragpath, cells = atac.cells)

### quantify multiome peaks in the scATAC-seq dataset
counts <- FeatureMatrix(
  fragments = atac.frags,
  features = granges(Aggr_Cd8a_pos),
  cells = atac.cells
)

Error in (function (classes, fdef, mtable) :
unable to find an inherited method for function ‘granges’ for signature ‘"Assay"’

**Reply by timoast** [reply]
Again, the `granges()` function will only work for a ChromatinAssay. Here you don't have that, you just have a standard Seurat assay. If this is a multiome object, you probably have an assay containing the ATAC data and an assay containing the RNA data. You need to run the `granges()` function on the assay containing the ATAC data.

**Reply by biobarani** [reply]
Hi Timoast,
Thanks for your help. I sort out that error in granges. its now works.

Now new error does not allow me to move forward.  I'm getting an error in FindintegrationAnchors function
**Error: Cannot find 'lsi' in this Seurat object**


For your reference, I have given the whole script. please help me with this.



counts <- Read10X_h5("~/Desktop/my_cellrangerARC_run/DW1_WT/filtered_feature_bc_matrix.h5")
fragpath <- "~/Desktop/my_cellrangerARC_run/DW1_WT/atac_fragments.tsv.gz"
fragcounts <- CountFragments(fragments = fragpath)
atac.cells <- fragcounts[fragcounts$frequency_count > 2000, "CB"]

atac.frags <- CreateFragmentObject(path = fragpath, cells = atac.cells)
annotation <- GetGRangesFromEnsDb(ensdb = EnsDb.Mmusculus.v79)
seqlevelsStyle(annotation) <- "UCSC"


pbmc <- CreateSeuratObject(
  counts = counts$`Gene Expression`,
  assay = "RNA"
)


pbmc.atac <- CreateChromatinAssay(
  counts = counts$Peaks,
  sep = c(":", "-"),
  fragments = fragpath,
  annotation = annotation
)

pbmc


counts <- FeatureMatrix(
  fragments = atac.frags,
  features = granges(pbmc.atac),
  cells = atac.cells
)


atac.assay <- CreateChromatinAssay(
  counts = counts,
  min.features = 1000,
  fragments = atac.frags
)
pbmc.atac <- CreateSeuratObject(counts = atac.assay, assay = "peaks")
pbmc.atac <- subset(pbmc.atac, nCount_peaks > 2000 & nCount_peaks < 30000)

pbmc.atac <- FindTopFeatures(pbmc.atac, min.cutoff = 10)
pbmc.atac <- RunTFIDF(pbmc.atac)
pbmc.atac <- RunSVD(pbmc.atac)


pbmc.atac$dataset <- "ATAC"
pbmc$dataset <- "Multiome"

pbmc.combined <- merge(pbmc.atac, pbmc)

pbmc.combined <- FindTopFeatures(pbmc.combined, min.cutoff = 10)
pbmc.combined <- RunTFIDF(pbmc.combined)
pbmc.combined <- RunSVD(pbmc.combined)

pbmc.combined <- RunUMAP(pbmc.combined, reduction = "lsi", dims = 2:30)
DimPlot(pbmc.combined, group.by = "dataset")
p1 <- DimPlot(pbmc.combined, group.by = "dataset")

**
integration.anchors <- FindIntegrationAnchors(
  object.list = list(pbmc, pbmc.atac),
  anchor.features = rownames(pbmc),
  reduction = "rlsi",
  dims = 2:30
)

**Error: Cannot find 'lsi' in this Seurat object**

In addition: Warning message:
In CheckDuplicateCellNames(object.list = object.list) :
  Some cell names are duplicated across objects provided. Renaming to enforce unique cell names.

Your help is really appreciated. Thanks.

**Reply by timoast** [reply]
According to the code above you did not compute LSI for the `pbmc` object

**Reply by biobarani** [reply]
Thank you very much for your quick reply,

Sorry, I'm following scATAC-seq data integration (https://satijalab.org/signac/articles/integrate_atac.html)
Based on the vignette combined ATAC and RNA object is used to compute LSI and I did the same but it still shows the error in FindIntegrationAnchors(). 

# merge
pbmc.combined <- merge(pbmc.atac, pbmc)

# process the combined dataset
pbmc.combined <- FindTopFeatures(pbmc.combined, min.cutoff = 10)
pbmc.combined <- RunTFIDF(pbmc.combined)
pbmc.combined <- RunSVD(pbmc.combined)

pbmc.combined <- RunUMAP(pbmc.combined, reduction = "lsi", dims = 2:30)

integration.anchors <- FindIntegrationAnchors(
  object.list = list(pbmc, pbmc.atac),
  anchor.features = rownames(pbmc),
  reduction = "rlsi",
  dims = 2:30
)

 Error: Cannot find 'lsi' in this Seurat object

If I run only for object containing RNA data it shows following error
pbmc.combined <- RunUMAP(pbmc, reduction = "lsi", dims = 2:30)

Error: Cannot find 'lsi' in this Seurat object

Thanks.

**Reply by timoast** [reply]
For reciprocal LSI projection (setting `reduction='rlsi'` in `FindIntegrationAnchors()`), you need to have an LSI dimension reduction present in each Seurat object to be integrated, so you need to compute one for the `pbmc` object also. In the integration vignette, we start with a pre-processed multiome object and so don't show all the steps again for processing that object, but you can see the steps here: https://satijalab.org/signac/articles/pbmc_multiomic.html

**Reply by biobarani** [reply]
Hi Timoast,


I will look into it and let you know if I have any difficulty. Thanks once again!
